### PR TITLE
Including jazz_is-slack-channel-available API and its configuration

### PR DESCRIPTION
**Pull request by supritat**
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### CODE REVIEW

File: core/jazz_is-slack-channel-available/components/config.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// =========================================================================
```
Comment (svidhani) [inline core/jazz_is-slack-channel-available/components/config.js]: you will have to use the new version
https://github.com/tmobile/jazz/blob/v1.7.1/core/jazz_usermanagement/components/config.js

File: core/jazz_is-slack-channel-available/index.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    if (event.method !== 'GET') {
+      reject({
+        result: "inputError",
+        message: "Unsupported method/request"
+      });
+    }
+
+    if (event.method === 'GET' && (!event.query || Object.keys(event.query).indexOf('slack_channel') === -1)) {
```
Comment (svidhani) [inline core/jazz_is-slack-channel-available/index.js]: includes?

Comment (svidhani) [reply core/jazz_is-slack-channel-available/index.js]: i guess the runtime will be 6.1?

File: core/jazz_is-slack-channel-available/index.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    var slack_token = "token=" + config.slack_channel_token;
+    //Pull endpoints from config file
+    var public_channel_url = config.public_channel_endpoint + slack_token;
+    var priv_channel_url = config.priv_channel_endpoint + slack_token;
+    var urlList = [];
+    urlList.push(get_response(public_channel_url, channel_name));
+    urlList.push(get_response(priv_channel_url, channel_name));
+    Promise.all(urlList)
+      .then(res => {
+        if (res.indexOf('true') !== -1) {
```
Comment (svidhani) [inline core/jazz_is-slack-channel-available/index.js]: i guess this is an array you can use includes

File: core/jazz_is-slack-channel-available/index.js [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+      })
+      .catch(error => {
+        logger.error("error: " + JSON.stringify(error));
+        reject(error);
+      });
+  });
+}
+
+function get_response(channel_url, channel_name) {
+  'use strict';
```
Comment (svidhani) [inline core/jazz_is-slack-channel-available/index.js]: why is this here. if your runtime is set to 6.1 then you dont need use strict. even if you have to use this statement this should be at the top of the module.

File: core/jazz_is-slack-channel-available/index.js [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  logger.info("Inside get_response:" + channel_url + ",channel_name:" + channel_name);
+  return new Promise((resolve, reject) => {
+    var params = {
+      method: 'GET',
+      uri: channel_url,
+      rejectUnauthorized: false
+    };
+    request(params, (error, response, body) => {
```
Comment (svidhani) [inline core/jazz_is-slack-channel-available/index.js]: new line?

File: core/jazz_is-slack-channel-available/index.js [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        var data = JSON.parse(response.body);
+        logger.info("Channels data: " + JSON.stringify(data));
+        if (data.ok) {
+          var list = data.channels ? data.channels : data.groups;
+          if (list.length) {
+            var count = 0;
+            for (var i in list) {
+              if (list[i].name === channel_name) {
+                resolve('true');
+                break;
```
Comment (svidhani) [inline core/jazz_is-slack-channel-available/index.js]: if you are resolving then why this?

File: core/jazz_is-slack-channel-available/index.js [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        logger.error(error);
+        reject(error);
+      } else {
+        var data = JSON.parse(response.body);
+        logger.info("Channels data: " + JSON.stringify(data));
+        if (data.ok) {
+          var list = data.channels ? data.channels : data.groups;
+          if (list.length) {
+            var count = 0;
+            for (var i in list) {
```
Comment (svidhani) [inline core/jazz_is-slack-channel-available/index.js]: let?

File: core/jazz_is-slack-channel-available/components/config.js (hunk 1 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/components/config.js]: Included new version config file

File: core/jazz_is-slack-channel-available/index.js (hunk 2 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: Using includes

File: core/jazz_is-slack-channel-available/index.js (hunk 3 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: using incldes

File: core/jazz_is-slack-channel-available/index.js (hunk 4 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: Removed 'use strict'.

File: core/jazz_is-slack-channel-available/index.js (hunk 5 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: included

File: core/jazz_is-slack-channel-available/index.js (hunk 6 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: Removed break.

File: core/jazz_is-slack-channel-available/index.js (hunk 7 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: using 'let'.

File: core/jazz_is-slack-channel-available/README.md [hunk 8]
```diff
@@ -1 +1,33 @@
+## Cloud API to check if a slack channel exists in T-Mobile Slack (t-mo.slack.com).
```
Comment (suryajak) [inline core/jazz_is-slack-channel-available/README.md]: Update description for OSS

File: core/jazz_is-slack-channel-available/index.js [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        if (error.result === 'inputError') {
+          return cb(JSON.stringify(errorHandler.throwInputValidationError(error.message)));
+        } else if (error.result === 'unauthorized') {
+          return cb(JSON.stringify(errorHandler.throwUnauthorizedError(error.message)));
+        } else {
+          cb(JSON.stringify(errorHandler.throwInternalServerError("Unhandled error.")));
+        }
+      });
+  } catch (e) {
+    cb(JSON.stringify(errorHandler.throwInternalServerError("Unhandled error.")));
```
Comment (suryajak) [inline core/jazz_is-slack-channel-available/index.js]: log error

File: core/jazz_is-slack-channel-available/index.js [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+  };
+
+  try {
+    genericInputValidation(event)
+      .then((res) => requestToChannels(config, resObj, res))
+      .then((res) => {
+        cb(null, responseObj(res, event.query));
+      })
+      .catch(error => {
+        if (error.result === 'inputError') {
```
Comment (suryajak) [inline core/jazz_is-slack-channel-available/index.js]: log error

File: core/jazz_is-slack-channel-available/index.js [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+          cb(JSON.stringify(errorHandler.throwInternalServerError("Unhandled error.")));
+        }
+      });
+  } catch (e) {
+    cb(JSON.stringify(errorHandler.throwInternalServerError("Unhandled error.")));
+  }
+};
+
+function genericInputValidation(event) {
+  logger.info("inside genericInputValidation");
```
Comment (suryajak) [inline core/jazz_is-slack-channel-available/index.js]: debug at best

File: core/jazz_is-slack-channel-available/index.js [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        message: "Unauthorized."
+      });
+    }
+
+    resolve(event.query.slack_channel);
+  });
+}
+
+function requestToChannels(config, resObj, channel_name) {
+  logger.info("Inside requestToChannels:" + channel_name);
```
Comment (suryajak) [inline core/jazz_is-slack-channel-available/index.js]: debug

File: core/jazz_is-slack-channel-available/index.js [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+        resolve(resObj);
+      })
+      .catch(error => {
+        logger.error("error: " + JSON.stringify(error));
+        reject(error);
+      });
+  });
+}
+
+function get_response(channel_url, channel_name) {
```
Comment (suryajak) [inline core/jazz_is-slack-channel-available/index.js]: use camelCase

File: core/jazz_is-slack-channel-available/index.js [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+  logger.info("Inside get_response:" + channel_url + ",channel_name:" + channel_name);
```
Comment (suryajak) [inline core/jazz_is-slack-channel-available/index.js]: debug

File: core/jazz_is-slack-channel-available/index.js [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+      if (error) {
+        logger.error(error);
+        reject(error);
+      } else {
+        var data = JSON.parse(response.body);
+        logger.info("Channels data: " + JSON.stringify(data));
+        if (data.ok) {
+          var list = data.channels ? data.channels : data.groups;
+          if (list.length) {
+            var count = 0;
```
Comment (suryajak) [inline core/jazz_is-slack-channel-available/index.js]: list.find for your code from 142-150

File: core/jazz_is-slack-channel-available/swagger/swagger.json [hunk 16]
```diff
@@ -0,0 +1,386 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "version": "2017-01-30T23:30:18Z",
+    "title": "{api_deployment_node_title}",
+    "description": "API specification for service that checks if a channel exists in t-mo.slack.com"
```
Comment (suryajak) [inline core/jazz_is-slack-channel-available/swagger/swagger.json]: Update description

File: core/jazz_is-slack-channel-available/components/error-handler.js [hunk 17]
```diff
@@ -0,0 +1,43 @@
+/**
+	Nodejs Template Project
```
Comment (svidhani) [inline core/jazz_is-slack-channel-available/components/error-handler.js]: header- please update header in all the files that are missing it.

File: core/jazz_is-slack-channel-available/README.md (hunk 8 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/README.md]: Updated

File: core/jazz_is-slack-channel-available/index.js (hunk 9 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: logging error

File: core/jazz_is-slack-channel-available/index.js (hunk 10 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: logging error

File: core/jazz_is-slack-channel-available/index.js (hunk 11 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: Done

File: core/jazz_is-slack-channel-available/index.js (hunk 12 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: Done

File: core/jazz_is-slack-channel-available/index.js (hunk 13 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: Done

File: core/jazz_is-slack-channel-available/index.js (hunk 14 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: Done

File: core/jazz_is-slack-channel-available/index.js (hunk 15 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: Done

File: core/jazz_is-slack-channel-available/swagger/swagger.json (hunk 16 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/swagger/swagger.json]: Updated

File: core/jazz_is-slack-channel-available/components/error-handler.js (hunk 17 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/components/error-handler.js]: Included header

File: core/jazz_is-slack-channel-available/index.js [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+            list.find(each =>{
+              if(each.name === channel_name) {
+                resolve('true');
+              } else {
+                count++;
+                if (count === list.length) {
+                  resolve('false');
```
Comment (svidhani) [inline core/jazz_is-slack-channel-available/index.js]: are you sure this works. if the first one in the loop goes to else then it will just resolve to false and return. I dont think this will ever resolve to true. Have you tested this?

File: core/jazz_is-slack-channel-available/index.js [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+  logger.info(event);
+
+  var resObj = {
+    'is_available': false
+  };
+
+  try {
+    genericInputValidation(event)
+      .then((res) => requestToChannels(config, resObj, res))
+      .then((res) => {
```
[the remark this answers is not in the document]
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: removed {}.

File: core/jazz_is-slack-channel-available/index.js (hunk 18 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: Yes it works. 
If the first one in the loop goes to else then it will check for the count value with list length based on that check it resolves with false and return. If count value is !== list.length, then it take the next one in the loop.

File: core/jazz_is-slack-channel-available/index.js [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+    request(params, (error, response, body) => {
+      if (error) {
+        logger.error(error);
+        reject(error);
+      } else {
+        var data = JSON.parse(response.body);
+        if (data.ok) {
+          var list = data.channels ? data.channels : data.groups;
+          if (list.length) {
+            var count = 0;
```
Comment (suryajak) [inline core/jazz_is-slack-channel-available/index.js]: Line 141 - 150 is equivalent to
```
if(list.find(a => a.name === channel_name)) {
      return resolve('true');
}
```

File: core/jazz_is-slack-channel-available/swagger/swagger.json [hunk 21]
```diff
@@ -0,0 +1,386 @@
...
+          "500": {
+            "description": "Internal Server Error",
+            "schema": {
+              "$ref": "#/definitions/500ErrorResponse"
+            }
+          }
+        }
+      },
+      "get": {
+        "description": "GET method to check if a channel exists in t-mo.slack.com",
```
Comment (suryajak) [inline core/jazz_is-slack-channel-available/swagger/swagger.json]: update description

File: core/jazz_is-slack-channel-available/index.js [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+    //Pull endpoints from config file
+    var public_channel_url = config.public_channel_endpoint + slack_token;
+    var priv_channel_url = config.priv_channel_endpoint + slack_token;
+    var urlList = [];
+    urlList.push(getResponse(public_channel_url, channel_name));
+    urlList.push(getResponse(priv_channel_url, channel_name));
+
+    Promise.all(urlList)
+      .then(res => {
+        if (res.includes('true')) {
```
Comment (suryajak) [inline core/jazz_is-slack-channel-available/index.js]: Line 109 - 113 is equivalent to:
```
resObj.is_available = res.includes('true');
```

File: core/jazz_is-slack-channel-available/index.js (hunk 20 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: Updated as per comment.

File: core/jazz_is-slack-channel-available/swagger/swagger.json (hunk 21 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/swagger/swagger.json]: Updated

File: core/jazz_is-slack-channel-available/index.js (hunk 22 above)
Comment (supritat) [reply core/jazz_is-slack-channel-available/index.js]: Updated as per comment.
